refactor!: improve/extend log settings

ankush opened the pull request:
closes #17014 

#### Before:

3 hardcoded doctypes:
<img width="1358" alt="Screenshot 2022-06-13 at 10 09 59 AM" src="https://user-images.githubusercontent.com/9079960/173280887-3be8d54e-0bcf-448e-9fda-ceb4c6ee566f.png">


#### After:

Table that can let user configure any doctype that supports log clearing. 

<img width="1312" alt="Screenshot 2022-06-13 at 10 08 36 AM" src="https://user-images.githubusercontent.com/9079960/173280914-a8bb21fb-3f11-420e-9252-c6ee855dbffe.png">


Any doctype that implements the defined interface automatically shows up on the list:

```python
@runtime_checkable
class LogType(Protocol):
  """Interface requirement for doctypes that can be cleared using log settings."""
  
  @staticmethod
  def clear_old_logs(days: int) -> None:
    ...
````


<img width="785" alt="Screenshot 2022-06-13 at 10 09 01 AM" src="https://user-images.githubusercontent.com/9079960/173280898-d285c247-d32e-47f9-b283-c7b645c122eb.png">


---

Show instructions on listview, thanks for the tip @gavindsouza 

<img width="913" alt="Screenshot 2022-06-14 at 1 39 07 PM" src="https://user-images.githubusercontent.com/9079960/173527232-124d8313-bea5-48ba-ba9c-902e51db0dab.png">

---

In addition to this also added a command line util to empty very large table that can't be cleared using DELETE queries because of the size. 


```sh
λ bench --site sitename clear-log-table --doctype="Error Log" --days=30
Backed up Error Log
Copying Error Log records from last 30 days to temporary table.
Cleared Error Log records older than 30 days
```


TODO:
- [x] patch for big deletes for users migrating to v14 
- [x] Message about retention setting on list / formview.

ankush committed:
fix!: dont auto set old logs as "seen"

Not sure why this is ever required, also not configurable.

ankush committed:
feat: log settings with "interface"

We have hardcoded "Log settings" to only apply on 3 doctypes, there are
few more logging doctypes in core which are not cleared right now, on
top of that it's not easy for user to configure all logging behaviour
from one place.

This change adds a table on log settings where logging doctypes that
support the interface required by log settings can auto-register and
show up in settings.

Currently only supported configuration is "number of days" to keep.

ankush committed:
feat: support log clearing for more doctypes

- Route History
- Error Snapshot
- Scheduled Job Log (only completed ones)

ankush committed:
feat: clear-log-table to clear large log tables

ankush committed:
refactor: slightly resilient patch and defaults

ankush commented on frappe/core/doctype/error_log/error_log.py:
```diff
@@ -12,13 +14,10 @@ def onload(self):
 			frappe.db.commit()
 
-
-def set_old_logs_as_seen():
```
I have removed this cause it made little sense to auto-mark old errors as seen, especially when they are auto-deleted anyway 😄

ankush commented on frappe/core/doctype/log_settings/log_settings.py:
```diff
@@ -2,63 +2,143 @@
-			else:
-				return _get_response(show_alert=False)
-		else:
-			return _get_response()
```
Removed all these conditions, in all the cases it did the same thing. It's just "notify everyone" with extra steps.

ankush commented on frappe/core/doctype/log_settings/log_settings.py:
```diff
@@ -2,63 +2,143 @@
+	"""Interface requirement for doctypes that can be cleared using log settings."""
+
+	@staticmethod
+	def clear_old_logs(days: int) -> None:
```
I went with just 1 config parameter for now, IDK if anything else is really "generic" or important. Doctypes can implement those if they wish (like "only clearing logs success status")

@gavindsouza  @phot0n

gavindsouza replied:
Or, only clearing "seen" logs 🤔

ankush replied:
@gavindsouza we can add filtering for statuses then?  👀 


That's common across most doctypes.

ankush commented on frappe/core/doctype/log_settings/log_settings.py:
```diff
@@ -2,63 +2,143 @@
+			# future addition that might happen to the required interface.
+			kwargs = frappe.get_newargs(func, {"days": entry.days})
+			func(**kwargs)
+			frappe.db.commit()
```
Committing by default makes sense, we are just clearing old garbage.

gavindsouza commented on frappe/commands/site.py:
```diff
@@ -1088,6 +1088,81 @@ def build_search_index(context):
+
+		original = f"`tab{doctype}`"
+		temporary = f"`tab{doctype} temp_table`"
+		backup = f"`tab{doctype} backup_table`"
```
```suggestion
		original = get_table_name(doctype)
		temporary = f"`{original} temp_table`"
		backup = f"`{original} backup_table`"
```

We can support __ tables too if we use `frappe.utils.get_table_name`

gavindsouza commented on frappe/core/doctype/log_settings/log_settings.py:
```diff
@@ -2,63 +2,143 @@
+from frappe.utils import cint
+from frappe.utils.caching import site_cache
+
+DEFAULT_LOGTYPES = {
```
Maybe add an identifier specifying "retention in days" var? 👅

gavindsouza commented on frappe/patches/v14_0/log_settings_migration.py:
```diff
@@ -0,0 +1,28 @@
+	try:
+		return frappe.db.get_single_value("Log Settings", fieldname)
+	except Exception:
+		pass
```
Might as well `frappe.db.get_singles_dict` - single query instead...But it makes very little difference 🍭

ankush replied:
get_sigles_dict is an ooof rn

ankush committed:
fix(ux): make log setting grid editable

ankush committed:
refactor: better varname and use get_table_name

Co-Authored-By: gavin <gavin18d@gmail.com>

ankush committed:
test: ensure temp tables are removed exist

ankush committed:
fix: clear pending logs using big-deletes

If some sites have long pending old data that doesn't get cleared
automcatically then this patch now attempts to discard old data by using
"big-delete" code instead of typical delete query.

ankush commented on frappe/core/doctype/log_settings/log_settings.py:
```diff
@@ -2,63 +2,188 @@
+	"Error Log": 30,
+	"Activity Log": 90,
+	"Email Queue": 30,
+	"Route History": 90,
```
I am unsure about this, should we add this by default? 👀 

I think it should be fine cause if activity log is okay to delete (which contains auth related logs) then route history should be fine too. 


@gavindsouza @surajshetty3416

gavindsouza replied:
Although it's unreliable, maybe let's not enable Route History deletion by default...Primarily because it'd be a breaking change 😅. Another way to go about this, is set a higher retention limit - say 1 year?

ankush replied:
Removing route history from defaults, users can add it back on their own if they want it.

ankush replied:
Error snapshot and scheduled job logs are fair game IMO, they are similar to error logs and aren't required for any sort of "forensics" after 90ish days.

gavindsouza commented:
Can we add a note to the list view of each of these logs saying `"{} are retained for {} days. Older records are automatically deleted. Click here to change Log Retention Settings"` - maybe add a link to enable log cleanup on other logs that don't have deletion enabled too.

ankush committed:
refactor: dont delete route history by default

It can still be added to log settings manually.

ankush committed:
feat(ux): show log retention policy in sidebar

for discoverablity

ankush commented:
Not fixing sider style issues. waste of time and contributes to unnecessary CI usage. Also half of them are false positive or unrelated to the diff 😑
